torch/_numpy/_ufuncs.py

```diff
@@ -97,7 +97,11 @@ def cast(x, dtype):
         elif isinstance(x1, torch.Tensor) and isinstance(x2, torch.Tensor):
             dtype = _dtypes_impl.result_type_impl(x1, x2)
             x1, x2 = _util.typecast_tensors((x1, x2), dtype, casting)
-        else:
+        elif not (
+            any(isinstance(x, torch.Tensor) for x in [x1, x2])
+            and any(_dtypes_impl.is_scalar(x) for x in [x1, x2])
+        ):
+            # If one is already scalar and the other is tensor, no need for casting the scalar.
```

Comment:
This is not right. NEP 50 is exactly to handle the tensor <> scalar case. A proper fix would go through checking in what cases we can avoid casting the scalar into a tensor.

No need to add a test, as this is thoroughly testesd in the codebase. In particular, the test you want to make sure that passes locally is `test/torch_np/test_nep50_examples.py` (and more generally, the whole `test/torch_np`)

Reply:
I see, thanks for the help, I will update the code!

Reply:
Hey @lezcano, the way I am trying to handle it now is inside is by doing a comparison between the weak and the not weak data types like this `weak_dtype!=dt`. The issue is that some of the NEP 50 functions like `arctan2` require tensors for both arguments. So, I am confused how to handle that without explicitly checking for functions, which doesn't seem right to do either. Any suggestions on how to proceed? Should I be fine tuning my checks a little more?

Reply:
Hi @lezcano , 
I'm currently comparing the weak and not weak data types using weak_dtype != dt. However, an issue is that certain NEP 50 functions, such as `arctan2`, which require tensors for both arguments. Thus I am uncertain about how to handle this situation without resorting to explicit function checks, which doesn't seem like the right approach either. Do you have any suggestions on how to proceed? Should I consider refining my checks further?

Reply:
That makes a lot of sense, and along the lines of what I was thinking, thank you for confirming!